packages/valory/skills/abstract_round_abci/utils.py

```diff
@@ -147,3 +167,302 @@ def parse_tendermint_p2p_url(url: str) -> Tuple[str, int]:
+        if tp is typing.Generic:
+            return typing.Generic
+        return None
+
+    def _get_args(tp):  # type: ignore
+        """Copied from the Python 3.8 typing module"""
+        if isinstance(tp, typing._GenericAlias):  # pylint: disable=protected-access
+            res = tp.__args__
+            if get_origin(tp) is collections.abc.Callable and res[0] is not Ellipsis:
+                res = (list(res[:-1]), res[-1])
+            return res
+        return ()
+
+
+def get_origin(tp):  # type: ignore
+    """
+    Get the unsubscripted version of a type.
+
+    This supports generic types, Callable, Tuple, Union, Literal, Final and
+    ClassVar. Returns None for unsupported types.
+    Examples:
+        get_origin(Literal[42]) is Literal
+        get_origin(int) is None
+        get_origin(ClassVar[int]) is ClassVar
+        get_origin(Generic) is Generic
+        get_origin(Generic[T]) is Generic
+        get_origin(Union[T, int]) is Union
+        get_origin(List[Tuple[T, T]][int]) == list
+    """
+    return _get_origin(tp)
+
+
+def get_args(tp):  # type: ignore
+    """
+    Get type arguments with all substitutions performed.
+
+    For unions, basic simplifications used by Union constructor are performed.
+    Examples:
+        get_args(Dict[str, int]) == (str, int)
+        get_args(int) == ()
+        get_args(Union[int, Union[T, int], str][int]) == (int, str)
+        get_args(Union[int, Tuple[T, int]][str]) == (int, Tuple[str, int])
+        get_args(Callable[[], T][int]) == ([], int)
+    """
+    return _get_args(tp)
+
+
+##
+# The following is borrowed from https://github.com/tamuhey/dataclass_utils/blob/81580d2c0c285081db06be02b4ecdd125532bef5/dataclass_utils/type_checker.py#L152
+##
+
+
+def is_pep604_union(ty: Type[Any]) -> bool:
+    """Check if a type is a PEP 604 union."""
+    return sys.version_info >= (3, 10) and ty is types.UnionType  # type: ignore # noqa: E721
+
+
+def _path_to_str(path: List[str]) -> str:
+    """Convert a path to a string."""
+    return " -> ".join(reversed(path))
+
+
+class AutonomyTypeError(TypeError):
+    """Type Error for the Autonomy type check system."""
+
+    def __init__(
+        self,
+        ty: Type[Any],
+        value: Any,
+        path: Optional[List[str]] = None,
+        exception: Optional[Any] = None,
+    ):
+        """Initialize AutonomyTypeError."""
+        self.ty = ty
+        self.value = value
+        self.path = path or []
+        self.exception = exception
+        super().__init__()
+
+    def __str__(self) -> str:
+        """Get string representation of AutonomyTypeError."""
+        path = _path_to_str(self.path)
+        msg = f"Error in field '{path}'. Expected type {self.ty}, got {type(self.value)} (value: {self.value})"
+        if self.exception is not None:
+            msg += f"\n{type(self.exception)}: {self.exception}"
+        return msg
+
+
+Result = Optional[AutonomyTypeError]  # returns error context
```

Comment:
Quite off topic, but have you considered monadic error handling? I think it can drastically reduce the complexity of our systems, considering the current control flow complexity via exception handling.

Reply:
Same comment as below